drivers/vdpa/virtio/virtio.h

```diff
@@ -0,0 +1,283 @@
+
+
+/*
+ * Maximum number of virtqueues per device.
+ */
+#define VIRTIO_MAX_VIRTQUEUE_PAIRS 8
+#define VIRTIO_MAX_VIRTQUEUES (VIRTIO_MAX_VIRTQUEUE_PAIRS * 2 + 1)
+
+/* VirtIO device IDs. */
+#define VIRTIO_ID_NETWORK  0x01
+#ifndef VIRTIO_ID_BLOCK
+#define VIRTIO_ID_BLOCK    0x02
+#endif
+#ifndef VIRTIO_ID_CONSOLE
+#define VIRTIO_ID_CONSOLE  0x03
+#endif
+#define VIRTIO_ID_ENTROPY  0x04
+#ifndef VIRTIO_ID_BALLOON
+#define VIRTIO_ID_BALLOON  0x05
+#endif
+#define VIRTIO_ID_IOMEMORY 0x06
+#ifndef VIRTIO_ID_9P
+#define VIRTIO_ID_9P       0x09
+#endif
+
+/* The bit of the ISR which indicates a device has an interrupt. */
+#define VIRTIO_ISR_INTR   0x1
+/* The bit of the ISR which indicates a device configuration change. */
+#define VIRTIO_ISR_CONFIG 0x2
+/* Vector value used to disable MSI for queue. */
+#define VIRTIO_MSI_NO_VECTOR 0xFFFF
+
+/* The alignment to use between consumer and producer parts of vring. */
+#define VIRTIO_VRING_ALIGN 4096
+
+/*
+ * This structure is just a reference to read net device specific
+ * config space; it is just a shadow structure.
+ *
+ */
+ #ifndef VIRTIO_API_FLAG
+struct virtio_net_config {
+	/* The config defining mac address (if VIRTIO_NET_F_MAC) */
+	uint8_t    mac[RTE_ETHER_ADDR_LEN];
+	/* See VIRTIO_NET_F_STATUS and VIRTIO_NET_S_* above */
+	uint16_t   status;
+	uint16_t   max_virtqueue_pairs;
+	uint16_t   mtu;
+	/*
+	 * speed, in units of 1Mb. All values 0 to INT_MAX are legal.
+	 * Any other value stands for unknown.
+	 */
+	uint32_t speed;
+	/*
+	 * 0x00 - half duplex
+	 * 0x01 - full duplex
+	 * Any other value stands for unknown.
+	 */
+	uint8_t duplex;
+	uint8_t rss_max_key_size;
+	uint16_t rss_max_indirection_table_length;
+	uint32_t supported_hash_types;
+} __rte_packed;
+#endif
+struct virtio_hw {
+	struct virtqueue **vqs;
+	uint64_t guest_features;
+	uint16_t vtnet_hdr_size;
+	uint8_t started;
+	uint8_t weak_barriers;
+	uint8_t vlan_strip;
+	bool rx_ol_scatter;
+	uint8_t has_tx_offload;
+	uint8_t has_rx_offload;
+	uint8_t use_vec_rx;
+	uint8_t use_vec_tx;
+	uint8_t use_inorder_rx;
+	uint8_t use_inorder_tx;
+	uint8_t opened;
+	uint16_t port_id;
+	uint8_t mac_addr[RTE_ETHER_ADDR_LEN];
+	/*
+	 * Speed is specified via 'speed' devarg or
+	 * negotiated via VIRTIO_NET_F_SPEED_DUPLEX
+	 */
+	bool get_speed_via_feat;
+	uint32_t speed;  /* link speed in MB */
+	uint8_t duplex;
+	uint8_t intr_lsc;
+	uint16_t max_mtu;
+	size_t max_rx_pkt_len;
+	/*
+	 * App management thread and virtio interrupt handler thread
+	 * both can change device state, this lock is meant to avoid
+	 * such a contention.
+	 */
+	rte_spinlock_t state_lock;
+	struct rte_mbuf **inject_pkts;
+	uint16_t max_queue_pairs;
+	uint16_t rss_rx_queues;
+	uint32_t rss_hash_types;
+	uint16_t *rss_reta;
+	uint8_t *rss_key;
+	uint64_t req_guest_features;
+	struct virtnet_ctl *cvq;
+	struct rte_pci_device *pci_dev;
+	const struct virtio_ops *virtio_ops;
+	struct rte_pci_ioport io;
+	bool use_va;
+};
+
+struct virtio_ops {
```

Comment:
is this code copied/modified from the existing dpdk code?
if so, can we use it in its original form?